resolver/resolvers.go

```diff
@@ -92,65 +98,60 @@ func createResolver(resolverURL, source string) (*Resolver, bool, error) {
 	}
 
 	switch u.Scheme {
-	case ServerTypeDNS, ServerTypeDoT, ServerTypeTCP:
+	case ServerTypeDNS, ServerTypeDoT, ServerTypeDoH, ServerTypeTCP:
+	case HttpsProtocol:
+		u.Scheme = ServerTypeDoH
 	default:
 		return nil, false, fmt.Errorf("DNS resolver scheme %q invalid", u.Scheme)
 	}
 
 	ip := net.ParseIP(u.Hostname())
-	if ip == nil {
+	isHostnameDomain := (ip == nil)
+	if ip == nil && u.Scheme != ServerTypeDoH {
 		return nil, false, fmt.Errorf("resolver IP %q invalid", u.Hostname())
 	}
 
 	// Add default port for scheme if it is missing.
-	var port uint16
-	hostPort := u.Port()
-	switch {
-	case hostPort != "":
-		parsedPort, err := strconv.ParseUint(hostPort, 10, 16)
-		if err != nil {
-			return nil, false, fmt.Errorf("resolver port %q invalid", u.Port())
-		}
-		port = uint16(parsedPort)
-	case u.Scheme == ServerTypeDNS, u.Scheme == ServerTypeTCP:
-		port = 53
-	case u.Scheme == ServerTypeDoH:
-		port = 443
-	case u.Scheme == ServerTypeDoT:
-		port = 853
-	default:
-		return nil, false, fmt.Errorf("missing port in %q", u.Host)
-	}
-
-	scope := netutils.GetIPScope(ip)
-	// Skip localhost resolvers from the OS, but not if configured.
-	if scope.IsLocalhost() && source == ServerSourceOperatingSystem {
-		return nil, true, nil // skip
+	port, err := parsePortFromURL(u)
+	if err != nil {
+		return nil, false, err
 	}
 
 	// Get parameters and check if keys exist.
 	query := u.Query()
-	for key := range query {
-		switch key {
-		case parameterName,
-			parameterVerify,
-			parameterBlockedIf,
-			parameterSearch,
-			parameterSearchOnly:
-			// Known key, continue.
-		default:
-			// Unknown key, abort.
-			return nil, false, fmt.Errorf(`unknown parameter "%s"`, key)
-		}
+	err = checkURLParameterValidity(u.Scheme, isHostnameDomain, query)
+	if err != nil {
+		return nil, false, err
 	}
 
 	// Check domain verification config.
+	serverAddress := ""
+	paramterServerIP := query.Get(parameterIP)
 	verifyDomain := query.Get(parameterVerify)
-	if verifyDomain != "" && u.Scheme != ServerTypeDoT {
-		return nil, false, fmt.Errorf("domain verification only supported in DOT")
+
+	if u.Scheme == ServerTypeDoT || u.Scheme == ServerTypeDoH {
+		switch {
+		case isHostnameDomain && paramterServerIP != "": // domain and ip as parameter
+			ip = net.ParseIP(paramterServerIP)
+			serverAddress = net.JoinHostPort(paramterServerIP, strconv.Itoa(int(port)))
+			verifyDomain = u.Hostname()
+		case !isHostnameDomain && verifyDomain != "": // ip and domain as parameter
+			serverAddress = net.JoinHostPort(ip.String(), strconv.Itoa(int(port)))
+		case isHostnameDomain && verifyDomain == "" && paramterServerIP == "": // only domain
+			verifyDomain = u.Hostname()
+		}
+	} else {
+		serverAddress = net.JoinHostPort(ip.String(), strconv.Itoa(int(port)))
+	}
```

Comment:
L132-145: Would it make sense to merge this with `checkURLParameterValidity` and have all options, parsing and validation in there?